python/__init__.py

```diff
@@ -0,0 +1,405 @@
+        return "cpuasm"
+
+    def get_architecture_descriptor(self, **kwargs):
+        # CPU does not have the following parameters, but we need to pass some values to
+        # make the Triton runtime happy.
+        return {"num_warps": 1, "num_stages": 1}
+
+    def make_launcher_stub(self, name, signature, constants, ids):
+        # name of files that are cached
+        so_cache_key = make_so_cache_key(version_key(), signature, constants, ids)
+        so_cache_manager = get_cache_manager(so_cache_key)
+        so_name = f"{name}.py"
+        # retrieve stub from cache if it exists
+        cache_path = so_cache_manager.get_file(so_name)
+        if cache_path is None:
+            kernel_placeholder_name = "KERNEL_NAME_PLACEHOLDER"
+            with tempfile.TemporaryDirectory() as tmpdir:
+                # Later KERNEL_NAME_PLACEHOLDER will be used to assign the kernel name
+                # in the following launch function.
+                launcher_src = _generate_launcher(constants, signature, kernel_placeholder_name)
+                # This function was renamed and made public in Python 3.10
+                if hasattr(sysconfig, 'get_default_scheme'):
+                    scheme = sysconfig.get_default_scheme()
+                else:
+                    scheme = sysconfig._get_default_scheme()
+                # 'posix_local' is a custom scheme on Debian. However, starting Python 3.10, the default install
+                # path changes to include 'local'. This change is required to use triton with system-wide python.
+                if scheme == 'posix_local':
+                    scheme = 'posix_prefix'
+                py_include_dir = sysconfig.get_paths(scheme=scheme)["include"]
+
+                dst_path = os.path.join(tmpdir, so_name)
+                py_src = f"""
+import os, subprocess, tempfile
+import importlib.util
+from pathlib import Path
+
+def launch(gridX, gridY, gridZ, num_warps, num_ctas, clusterDim0, clusterDim1, clusterDim2,
+           shared, stream, cu_function, launch_enter_hook, launch_exit_hook, compiled_kernel,
+           *args):
+    # Unlike CUDA/HIP, we cannot easily pass function pointer across different pybind libraries.
+    # Let's compile a kernel every time.
+    asm_src = compiled_kernel.asm["{self.get_kernel_bin()}"]
+    launcher_src = '''
+{launcher_src}
+'''.replace("{kernel_placeholder_name}", compiled_kernel.metadata["name"])
+    with tempfile.TemporaryDirectory() as tmpdir:
+        asm_src_path = os.path.join(tmpdir, "kernel.s")
+        launcher_src_path = os.path.join(tmpdir, "main.cxx")
+        so_path = os.path.join(tmpdir, "kernel.so")
+        Path(asm_src_path).write_text(asm_src)
+        Path(launcher_src_path).write_text(launcher_src)
+        # Compile it together.
+        ret = subprocess.check_call(["g++", launcher_src_path, asm_src_path, f"-I{py_include_dir}", f"-I{Path(__file__).resolve().parent}", "-shared", "-fPIC", "-o", so_path])
+        if ret != 0:
```

Comment:
nit: similar to the assert comment above, this is not needed

Reply:
Updated!